fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/properties/FDBPropertyKey.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * FDBPropertyKey.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.properties;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.function.Supplier;
+
+@SuppressWarnings("unused")
+/**
+ * A key for a property that adopter configures for Record Layer.
+ * The value is lazily loaded from the supplier when the property is read.
+ * If no supplier is provided, a default value is used.
+ */
+public final class FDBPropertyKey<T> {
```

Comment:
One bike-shedding comment on naming (hopefully the only one): I'm not sure this is really best named a `FDBPropertyKey` given that it's, well, not really defined in `FDB`. The name `RecordLayerPropertyKey` is a bit more accurate, or even just `PropertyKey` (and relying on the `package` to prevent us from potential collisions with other classes of that name). I don't have strong opinions here, but I thought I'd mention it.

Whatever convention is chosen, the `FDBRecordContextProperties` should probably follow suit.